src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var action = await IntroduceParameterAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (action != null)
+            {
+                context.RegisterRefactoring(action, textSpan);
+            }
+        }
+
+        public async Task<Solution> IntroduceParameterAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, bool trampoline, bool overload, CancellationToken cancellationToken)
+        {
+            if (trampoline || overload)
+            {
+                return await IntroduceParameterForTrampolineAsync(document, expression, allOccurrences, trampoline, cancellationToken).ConfigureAwait(false);
+            }
+            return await IntroduceParameterForRefactoringAsync(document, expression, allOccurrences, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<CodeAction?> IntroduceParameterAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var expression = await document.TryGetRelevantNodeAsync<TExpressionSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+            if (expression == null || CodeRefactoringHelpers.IsNodeUnderselected(expression, textSpan))
+            {
+                return null;
+            }
+
+            var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            var (title, actions) = AddActions(semanticDocument, expression, cancellationToken);
+
+            if (actions.Length > 0)
+            {
+                return new CodeActionWithNestedActions(title, actions, isInlinable: true);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForRefactoringAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            var updatedCallSitesSolution = await RewriteCallSitesAsync(document, expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
```

Comment:
this seems to indicate that it is only rewriting callsites.  sounds like it's also adding the parameter?

Reply:
Right, the name needs to be changed since it was all consolidated in that file to be able to do everything with the same editor.